compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs

```diff
@@ -611,6 +623,20 @@ impl<'brillig> Context<'brillig> {
     }
 }
 
+fn instruction_can_be_hoisted(
+    instruction: &Instruction,
+    dfg: &mut DataFlowGraph,
+    deduplicate_with_predicate: bool,
+) -> bool {
+    // These two can never be hoisted as they have a side-effect
+    // (though it's fine to de-duplicate them, just not fine to hoist them)
+    if matches!(instruction, Instruction::Constrain(..) | Instruction::RangeCheck { .. }) {
```

Comment:
Wouldn't it be enough to do this for `Constrain`, because that's the one that interacts with simplifications that can replace other instructions' results? Other instructions can have side effects too, but `Constrain` has special role in `constant_folding`. I don't see `RangeCheck` have the same.

Reply:
I think any instruction can be simplified. At least `instruction` here comes from `resolve_instruction` where any `Instruction` type can be returned.

Reply:
What I meant was that we only have simplifications that [come from Constrain](https://github.com/noir-lang/noir/blob/3306f490bf100899cf46f366be2afe3732054f35/compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs#L405), and I thought the hoisting of `Constrain` in the dominator is what's causing trouble further on. Just not sure what `RangeCheck` would do.